fetcher/fetcher.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fetcher
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+
+	"github.com/ava-labs/hypersdk/keys"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+// Fetcher retrieves values on-the-fly and
+// ensures that a value is only fetched from
+// disk once. Subsequent requests can be retrieved
+// from cache.
+type Fetcher struct {
+	im        state.Immutable
+	cacheLock sync.RWMutex
+	Cache     map[string]*FetchData
+
+	TxnsToFetch map[ids.ID]*sync.WaitGroup // Number of keys a txn is waiting on
+
+	stopOnce  sync.Once
+	stop      chan struct{}
+	err       error
+	fetchable chan *task
+	l         sync.Mutex
+
+	completed int
+	totalTxns int
+}
+
+// Data to insert into the cache
+type FetchData struct {
+	Val    []byte
+	Exists bool
+
+	Chunks uint16
+}
+
+// task holds the information that a worker needs to fetch values
+type task struct {
+	ctx      context.Context
+	id       ids.ID
+	toLookup []string
+}
+
+// New creates a new [Fetcher]
+func New(numTxs int, concurrency int, im state.Immutable) *Fetcher {
+	f := &Fetcher{
+		im:          im,
+		Cache:       make(map[string]*FetchData, numTxs),		
+		TxnsToFetch: make(map[ids.ID]*sync.WaitGroup, numTxs),
+		fetchable:   make(chan *task),
+		stop:        make(chan struct{}), // TODO: implement Stop()
+		totalTxns:   numTxs,
+	}
+	for i := 0; i < concurrency; i++ {
+		f.createWorker()
+	}
+	return f
+}
+
+func (f *Fetcher) createWorker() {
+	go func() {
+		for {
+			select {
+			case t, ok := <-f.fetchable:
+				if !ok {
+					return
+				}
+				for _, k := range t.toLookup {
+					// Allow concurrent reads to cache
+					f.cacheLock.RLock()
+					if _, ok := f.Cache[k]; ok {
+						f.TxnsToFetch[t.id].Done()
```

Comment:
You need to synchronize access to `TxnsToFetch` otherwise it could panic (you are reading a value here while another thread may add an id to it)